### PR TITLE
Logging

### PR DESCRIPTION
**Pull request by joshmeranda**
**Problem:**
Implement the logging feature as described in [577](https://github.com/harvester/harvester/issues/577).

**Solution:**
Installer the `rancher-logging` and `rancher-logging-crd` charts as `ManagedCharts` by default and route logs to local loki deployment using the logging operator from BonzaiCloud.

**Related Issue:**
https://github.com/harvester/harvester/issues/577
https://github.com/harvester/harvester/issues/2644
https://github.com/harvester/harvester/issues/2645
https://github.com/harvester/harvester/issues/2646
https://github.com/harvester/harvester/issues/2647

**Test plan:**

**Blocked By**
https://github.com/harvester/harvester/pull/2681

cc @guangbochen @bk201 @w13915984028

### CODE REVIEW

File: pkg/config/cos.go [hunk 1]
```diff
@@ -171,6 +172,10 @@ func initRancherdStage(config *HarvesterConfig, stage *yipSchema.Stage) error {
 		config.MonitoringChartVersion = MonitoringChartVersion
 	}
 
+	if config.LoggingChartVersion == "" {
+		config.MonitoringChartVersion = LoggingChartVersion
```
Comment (w13915984028) [inline pkg/config/cos.go:176]: config.MonitoringChartVersion ->  config.LoggingChartVersion

File: pkg/config/cos.go [hunk 2]
```diff
@@ -491,7 +496,16 @@ func genBootstrapResources(config *HarvesterConfig) (map[string]string, error) {
 	for _, templateName := range []string{
 		"10-harvester.yaml",
 		"11-monitoring-crd.yaml",
+
 		"13-monitoring.yaml",
+
+		"14-loki-stack.yaml",
```
Comment (w13915984028) [inline pkg/config/cos.go:502]: use different ID, 14, 15, 16...

File: pkg/config/templates/rancherd-14-logging-crd.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+- apiVersion: management.cattle.io/v3
+  kind: ManagedChart
+  metadata:
+    name: rancher-logging-crd
+    namespace: fleet-local
+  spec:
+    chart: rancher-logging-crd
+    releaseName: rancher-logging-crd
+    version: {{ .LoggingChartVersion }}
+    defaultNamespace: cattle-monitoring-system
```
Comment (w13915984028) [inline pkg/config/templates/rancherd-14-logging-crd.yaml:11]: defaultNamespace: cattle-monitoring-system -> cattle-logging-system

File: pkg/config/templates/rancherd-14-loki-stack.yaml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    kind: ManagedChart
+    metadata:
+      name: loki-stack
+      namespace: fleet-local
+    spec:
+      chart: loki-stack
+      releaseName: loki-stack
+      #      we might want to add this to the cos config eventually, so we can change the version with ldflags
+      version: 2.7.0
+      defaultNamespace: cattle-monitoring-system
```
Comment (w13915984028) [inline pkg/config/templates/rancherd-14-loki-stack.yaml:12]: defaultNamespace: cattle-monitoring-system  ??

File: pkg/config/templates/rancherd-15-logging.yaml [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+- apiVersion: management.cattle.io/v3
+  kind: ManagedChart
+  metadata:
+    name: rancher-logging
+    namespace: fleet-local
+  spec:
+    chart: rancher-logging
+    releaseName: rancher-logging
+    version: {{ .LoggingChartVersion }}
+    defaultNamespace: cattle-monitoring-system
```
Comment (w13915984028) [inline pkg/config/templates/rancherd-15-logging.yaml:11]: defaultNamespace: cattle-monitoring-system ??

File: pkg/config/templates/rancherd-15-logging.yaml [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+      - clusterName: local
+        clusterSelector:
+          matchExpressions:
+            - key: provisioning.cattle.io/unmanaged-system-agent
+              operator: DoesNotExist
+    values:
+      fluentd:
+        fluentdPvcSpec:
+          storageClassName: longhorn
+      alertmanager:
```
Comment (w13915984028) [inline pkg/config/templates/rancherd-15-logging.yaml:35]: is this file copied from `rancherd-13-monitoring`, there should no `alertmanager`, `grafana`

**Review by w13915984028 — COMMENTED** [review]
Needs further update.

File: pkg/config/templates/rancherd-15-logging.yaml (hunk 5 above)
Comment (joshmeranda) [reply pkg/config/templates/rancherd-15-logging.yaml]: changed to `cattle-logging-system`

File: pkg/config/templates/rancherd-14-loki-stack.yaml (hunk 4 above)
Comment (joshmeranda) [reply pkg/config/templates/rancherd-14-loki-stack.yaml]: changed to `cattle-logging-system`

File: pkg/config/templates/rancherd-14-logging-crd.yaml (hunk 3 above)
Comment (joshmeranda) [reply pkg/config/templates/rancherd-14-logging-crd.yaml]: changed to `cattle-logging-system`

File: pkg/config/cos.go (hunk 1 above)
Comment (joshmeranda) [reply pkg/config/cos.go]: fixed typo

File: pkg/config/templates/rancherd-15-logging.yaml (hunk 6 above)
Comment (joshmeranda) [reply pkg/config/templates/rancherd-15-logging.yaml]: removed the unnecessary values from `rancherd-15-logging.yaml`

File: scripts/version-logging [hunk 7]
```diff
@@ -0,0 +1,2 @@
+LOGGING_VERSION=100.1.2+up3.17.7
```
Comment (w13915984028) [inline scripts/version-logging:1]: Please double check, upstream rancher has `100.1.2+up3.17.4` and `100.1.3+up3.17.7`

https://github.com/rancher/charts/tree/dev-v2.6/charts/rancher-logging/100.1.2%2Bup3.17.4
https://github.com/rancher/charts/tree/dev-v2.6/charts/rancher-logging/100.1.3%2Bup3.17.7

File: scripts/build-bundle [hunk 8]
```diff
@@ -27,7 +27,8 @@ mkdir -p ${RANCHERD_IMAGES_DIR}
 harvester_path=../harvester
 if [ ! -d ${harvester_path} ];then
     echo "No existed harvester source. Pulling into /tmp/harvester"
-    git clone --branch master --single-branch --depth 1 https://github.com/harvester/harvester.git /tmp/harvester
+#    git clone --branch master --single-branch --depth 1 https://github.com/harvester/harvester.git /tmp/harvester
+    git clone --branch logging-chart --single-branch --depth 1 https://github.com/joshmeranda/harvester.git /tmp/harvester
```
Comment (w13915984028) [inline scripts/build-bundle:31]: personal git needs to be replaced

File: scripts/patch-harvester [hunk 9]
```diff
@@ -34,6 +34,6 @@ yq e -i '.webhook.image.tag = "'"${IMAGE_PUSH_TAG}"'"' ./deploy/charts/harvester
 yq e -i '.upgrade.image.tag = "'"${IMAGE_PUSH_TAG}"'"' ./deploy/charts/harvester/values.yaml
 
 echo "Replace the Harvester image repo to ${REPO}"
-replace_harvester_image_repo '.containers.apiserver.image.repository'
-replace_harvester_image_repo '.webhook.image.repository'
-replace_harvester_image_repo '.upgrade.image.repository'
\ No newline at end of file
+#replace_harvester_image_repo '.containers.apiserver.image.repository'
```
Comment (w13915984028) [inline scripts/patch-harvester:37]: need to revert changes

File: pkg/config/templates/rancherd-15-logging.yaml [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    version: {{ .LoggingChartVersion }}
+    defaultNamespace: cattle-logging-system
+    repoName: harvester-charts
+    diff:
+      comparePatches:
+        - apiVersion: admissionregistration.k8s.io/v1
+          kind: ValidatingWebhookConfiguration
+          name: rancher-logging-admission
+          operations:
+            - { "op": "remove", "path": "/webhooks" }
```
Comment (w13915984028) [inline pkg/config/templates/rancherd-15-logging.yaml:19]: The `diff` part is a special patch for `rancher-monitoring`, `rancher-logging` does not need it.

File: scripts/version-logging (hunk 7 above)
Comment (joshmeranda) [reply scripts/version-logging:1]: Yes both are in the rancher helm repo as of rancher 2.6.7

File: pkg/config/cos.go (hunk 2 above)
Comment (joshmeranda) [reply pkg/config/cos.go]: moved to `rancherd-16-loki-stack-yaml`

File: pkg/config/templates/rancherd-15-logging.yaml (hunk 10 above)
Comment (joshmeranda) [reply pkg/config/templates/rancherd-15-logging.yaml]: removed

File: scripts/patch-harvester (hunk 9 above)
Comment (joshmeranda) [reply scripts/patch-harvester]: reverted

File: scripts/build-bundle (hunk 8 above)
Comment (joshmeranda) [reply scripts/build-bundle]: replaced

File: pkg/config/templates/rancherd-15-logging.yaml (hunk 10 above)
Comment (joshmeranda) [reply pkg/config/templates/rancherd-15-logging.yaml]: removed

File: scripts/patch-harvester (hunk 9 above)
Comment (joshmeranda) [reply scripts/patch-harvester]: reverted

File: scripts/build-bundle (hunk 8 above)
Comment (joshmeranda) [reply scripts/build-bundle]: replaced

**Comment by w13915984028** [comment]
@joshmeranda 
Please check those 2 files, add `rancher-logging` related images into those two, and make sure the related images are pulled and packed into ISO, which is pre-condition for air-gapped environment.

```
scripts/images/allow.yaml
scripts/images/rancher-images.txt
```

**Comment by w13915984028** [comment]
Replaced by https://github.com/harvester/harvester-installer/pull/331.

Code is rebased and updated. @joshmeranda

**Pull request closed without merging**